clang/include/clang/CIR/Dialect/IR/CIRAttrs.td

```diff
@@ -386,4 +386,34 @@ def ASTFunctionDeclAttr : ASTDecl<"FunctionDecl", "fndecl">;
 def ASTVarDeclAttr : ASTDecl<"VarDecl", "vardecl">;
 def ASTRecordDeclAttr : ASTDecl<"RecordDecl", "recdecl">;
 
+
+def NoInline : I32EnumAttrCase<"NoInline", 1, "noInline">;
+def AlwaysInline : I32EnumAttrCase<"AlwaysInline", 2, "alwaysInline">;
+def InlineHint : I32EnumAttrCase<"InlineHint", 3, "inlineHint">;
+
+def InlineEnum : I32EnumAttr<"InlineEnum", "InlineEnum", [
+  NoInline, AlwaysInline, InlineHint
+]> {
+  let cppNamespace = "::mlir::cir";
+}
+
+def InlineAttr : CIR_Attr<"Inline", "inline"> {
```

Comment:
Right now this attribute is not tied to `FuncOp`s. Please add this attribute as an optional input to `FuncOp`, and use the printer to get a prettier output than `attributes { #cir.func = ...`, the printer output could be as simple as `inline(no)`, `inline(always)` and `inline(hint)`. (See `alias(...)` as an example).

Reply:
For now the attribute wouldn't be added to a function that doesn't need an inline attribute, so it's already optional. This is ensured by various checks in `CIRGenModule::setCodegenAttributesForFunc`.

Sounds good to simply the enum names to "no"m "always" and "hint".

Reply:
> For now the attribute wouldn't be added to a function that doesn't need an inline attribute, so it's already optional

IMO it's better to be explicit with what a `FuncOp` can accept, otherwise the attribute is logically loose and not obvious in tablegen where its used, passing it as an input also allows you to use a automatically generated getter and setter to deal with the attribute from a given `FuncOp`, instead of having to do the adhoc string query directly.

Reply:
I see your point and I think that'd be better to be applied to the outer attribute, i.e, cir.funcAttr which is  a set of optnone, inline .... attributes. Here is how it's set in `CIRGenModule::setCodegenAttributesForFunc`:
```

if (!attrs.empty())
    f->setAttr("cir.funcAttr",
               mlir::ArrayAttr::get(builder.getContext(), attrs));
```

But I'm still a bit confused that how to make it literally optional. cir.funcAttr is added to cir.funop's underlying optional attrs field after the funop is created. So it is not a direct input to the funcop constructor.

Reply:
Do you want to change the funcOp signature by allowing an optional funcattr?


  ```
let arguments = (ins SymbolNameAttr:$sym_name,
                       TypeAttrOf<CIR_FuncType>:$function_type,
                       UnitAttr:$builtin,
                       UnitAttr:$coroutine,
                       UnitAttr:$lambda,
                       UnitAttr:$no_proto,
                       DefaultValuedAttr<GlobalLinkageKind,
                                         "GlobalLinkageKind::ExternalLinkage">:$linkage,
                       OptionalAttr<StrAttr>:$sym_visibility,
                       OptionalAttr<DictArrayAttr>:$arg_attrs,
                       OptionalAttr<DictArrayAttr>:$res_attrs,
                       OptionalAttr<FlatSymbolRefAttr>:$aliasee,
                       OptionalAttr<ASTFunctionDeclAttr>:$ast);
```

That's what I started with originally, but then I realized we don't give anything at its construction time though, and later on we'll still need optionally set the attribute.